proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,329 @@
+
+    * See the "Lexer implementation" section below
+
+#. Add a new handler in ``rnExpr``, probably near the `OverloadedStrings implementation <https://gitlab.haskell.org/ghc/ghc/-/blob/a1e42e7ac6121404afb2a42e11d0c778ce0fe483/compiler/GHC/Rename/Expr.hs#L370>`_, to replace ``HsInterpolatedString`` with the corresponding expression.
+
+    * See the "Renamer implementation" section below
+
+Lexer implementation
+~~~~~~~~~~~~~~~~~~~~
+
+TODO: lex things inside ``${...}``
+TODO: ITstring vs ITmultilinestring
+
+Some examples:
+
+::
+  s"foo ${f a b} bar ${g x} baz ${name}"
+  ====>
+  HsInterpolatedString
+    (<SourceText>, False)
+    "foo "
+    [ (HsApp (HsApp (HsVar "f") (HsVar "a")) (HsVar "b"), " bar ")
+    , (HsApp (HsVar "g") (HsVar "x"), " baz ")
+    , (HsVar "name", "")
+    ]
+
+  s"foo ${bar}"
+  ====>
+  HsInterpolatedString
+    (<SourceText>, False)
+    "foo "
+    [(HsVar "bar", "")]
+
+  s"${foo} bar"
+  ====>
+  HsInterpolatedString
+    (<SourceText>, False)
+    ""
+    [(HsVar "foo", " bar")]
+
+  s"""
+  line 1 ${foo}
+  line ${foo} 2
+  """
+  ====>
+  HsInterpolatedString
+    (<SourceText>, True)
+    "line 1 "
+    [ (HsVar "foo", "\nline ")
+    , (HsVar "foo", " 2")
+    ]
+
+We will always require interpolation to be of the form ``${foo}``, unlike other languages where ``$foo`` is allowed for single-variable interpolation. This simplifies the syntax and avoids the need for escaping bare ``$``. Interpolation may be escaped by escaping the dollar sign; e.g. ``"\${foo}"`` is equivalent to ``['$', '{', 'f', 'o', 'o', '}']``.
+
+Any Haskell expression may go inside the braces, but an unmatched ``}`` will immediately end processing. This restriction simplifies the implementation and isn't a big deal in practice, since one could always break it out into a helper variable. This implies that nested string interpolation is not allowed.
+
+Renamer implementation
+~~~~~~~~~~~~~~~~~~~~~~
+
+After parsing, ``rnExpr`` gets something like:
+
+::
+  HsInterpolatedString
+    (<SourceText>, False)
+    "foo "
+    [ (<f a b>, " bar ")
+    , (<g x>  , " baz ")
+    , (<name> , "")
+    ]
+
+``rnExpr`` will transform this into the equivalent ``HsExpr`` for the following code:
+
+::
+  fromBuilder $
+    fromString "foo "
+    <> interpolate (f a b) <> fromString " bar "
+    <> interpolate (g x)   <> fromString " baz "
+    <> interpolate name    <> fromString ""
+
+Interpolation design
+~~~~~~~~~~~~~~~~~~~~
+
+The following type classes would be implemented in ``ghc-experimental`` under ``Data.String.Interpolate``. After the API has stablized, these would eventually live in ``GHC.Exts`` alongside ``IsString``.
+
+::
+
+  class (IsString (Builder s), Monoid (Builder s)) => Buildable s where
+    type Builder s :: Type
+    fromBuilder :: Builder s -> s
```

Comment:
A thing I don't think has been mentioned yet, and that seems worth mentioning:

This means that `s"foo" :: SomeType` can be very different than `"foo" :: SomeType`, despite not doing any interpolation. One or the other might fail to compile, and if they both compile then they might do different things.

If we don't want that we could suggest a law (given `IsString s`, `fromBuilder . fromString === fromString`).

Reply:
I'm imagining using this to make data structure construction prettier. E.g. if you have

```haskell
data Rose a = Node a [Rose a]
myRose =
  Node 1
    [ Node 11 []
    , Node 12
        [ Node 121 [] ]
    , Node 13 []
    ]
```

Maybe there's some builder type that lets you write it as

```haskell
myRose = s"""
    ${1}
    - ${11}
    - ${12}
      - ${121}
    - ${13}
"""

Is it good that you can do that kind of thing? Not sure. I don't think this specific example is very helpful (if you want this you can probably also implement some monad that lets you use do notation), but maybe there's something it would be good for.

Reply:
> This means that s"foo" :: SomeType can be very different than "foo" :: SomeType

I don't really think so. Note that there is no `IsString (Builder s)` instance involved here. We get
```
s"foo" 
=== {- desugar -}
fromBuilder (toBuilder (fromString "foo"))
=== {- builder iso -}
fromString "foo"
=== {- OverloadedStrings -}
"foo"
```

In https://github.com/ghc-proposals/ghc-proposals/pull/570#issuecomment-2268795638 I proposed to require `fromString @(Builder s) === toBuilder . fromString @s` which is equivalent to your requirement, but we do not really need that because we do not require the `IsString (Builder s)` instance.

> I'm imagining using this to make data structure construction prettier. 

Indeed we could think about nice syntax for `Buildable`. IMO we should not necessarily abuse string interpolation for that, though, so this is entirely separate from this proposal.

You could take inspiration in `blaze-html` and others and define a indentation-sensitive DSL based on do-notation:

```hs
myRose = do
  n 1 do
    n 11 []
    n 12 do
       n 121 []
    n 13 []
```

Reply:
I'm not sure how the tree structure arises. Can someone explain?

To take your example:

```haskell
myRose = s"""
    ${1}
    - ${11}
    - ${12}
      - ${121}
    - ${13}
"""
```

This would currently translate into

```hs
fromBuilder $
  toBuilder (fromString "")
  <> interpolate 1 <> toBuilder (fromString "\n- ")
  <> interpolate 11 <> toBuilder (fromString "\n- ")
  <> interpolate 12 <> toBuilder (fromString "\n  - ")
  <> interpolate 121 <> toBuilder (fromString "\n- ")
  <> interpolate 13 <> toBuilder (fromString "")
```
To get a tree structure, the interpolation would have to parse the indentation and recognize that deeper indentation is on a deeper level of the tree. I don't think this should be within the scope of string interpolation

Reply:
Ah, maybe the proposal just hasn't been updated for some of the discussion that's happened? (It can be hard to keep track.)

To be clear, [currently](https://github.com/ghc-proposals/ghc-proposals/blob/46f44928cedc93ed0f44cf92c8044f633b15ff3f/proposals/0570-string-interpolation.rst#22renamer-implementation) (as of 46f44928) the proposal would desugar `s"foo"` as `fromBuilder $ fromString "foo"`, not as `fromBuilder $ toBuilder $ fromString "foo"`.

If that's on the todo list to change, then fair enough. If we add `toBuilder`, then as long as `fromBuilder . toBuilder === id`, we'll have `s"foo" === "foo"`.

Reply:
Yes, I just woke up 🙂 So I haven't made that change yet. But yes, that is the expected behavior